consensus/istanbul/backend/api.go

```diff
@@ -35,66 +36,78 @@ type API struct {
+		head := api.chain.CurrentHeader()
+		if head == nil {
+			return nil, errUnknownBlock
+		}
+		if number == nil || *number == rpc.LatestBlockNumber {
+			parent = head.Number.Uint64() - 1
+		} else {
+			parent = head.Number.Uint64()
+		}
+	} else if *number == rpc.EarliestBlockNumber {
 		return nil, errUnknownBlock
+	} else {
+		parent = uint64(*number - 1)
 	}
-	return api.istanbul.snapshot(api.chain, header.Number.Uint64(), header.Hash(), nil)
-}
 
-// GetSnapshotAtHash retrieves the state snapshot at a given block.
-func (api *API) GetSnapshotAtHash(hash common.Hash) (*Snapshot, error) {
-	header := api.chain.GetHeaderByHash(hash)
+	header := api.chain.GetHeaderByNumber(parent)
 	if header == nil {
 		return nil, errUnknownBlock
 	}
-	return api.istanbul.snapshot(api.chain, header.Number.Uint64(), header.Hash(), nil)
+	return header, nil
 }
 
-// GetValidators retrieves the list of authorized validators at the specified block.
-func (api *API) GetValidators(number *rpc.BlockNumber) ([]common.Address, error) {
+// GetSnapshot retrieves the state snapshot at a given block.
+func (api *API) GetSnapshot(number *rpc.BlockNumber) (*Snapshot, error) {
 	// Retrieve the requested block number (or current if none requested)
 	var header *types.Header
 	if number == nil || *number == rpc.LatestBlockNumber {
 		header = api.chain.CurrentHeader()
 	} else {
 		header = api.chain.GetHeaderByNumber(uint64(number.Int64()))
 	}
-	// Ensure we have an actually valid block and return the validators from its snapshot
+	// Ensure we have an actually valid block and return its snapshot
 	if header == nil {
 		return nil, errUnknownBlock
 	}
-	snap, err := api.istanbul.snapshot(api.chain, header.Number.Uint64(), header.Hash(), nil)
+	return api.istanbul.snapshot(api.chain, header.Number.Uint64(), header.Hash(), nil)
+}
+
+// GetValidators retrieves the list validators that must sign a given block.
+func (api *API) GetValidators(number *rpc.BlockNumber) ([]common.Address, error) {
+	header, err := api.getParentHeaderByNumber(number)
 	if err != nil {
 		return nil, err
 	}
-	validators := snap.validators()
-	validatorsAddresses, _ := istanbul.SeparateValidatorDataIntoIstanbulExtra(validators)
-	return validatorsAddresses, nil
+	validators := api.istanbul.GetValidators(header.Number, header.Hash())
+	return istanbul.GetAddressesFromValidatorList(validators), nil
 }
 
-// GetValidatorsAtHash retrieves the state snapshot at a given block.
-func (api *API) GetValidatorsAtHash(hash common.Hash) ([]common.Address, error) {
-	header := api.chain.GetHeaderByHash(hash)
-	if header == nil {
-		return nil, errUnknownBlock
+// GetProposer retrieves the proposer for a given block number (i.e. sequence) and round.
+func (api *API) GetProposer(sequence *rpc.BlockNumber, round *uint64) (common.Address, error) {
+	header, err := api.getParentHeaderByNumber(sequence)
+	if err != nil {
+		return common.Address{}, err
```

Comment:
there's `common.ZeroAddress`. I'm never sure what would be a good null value...
i guess ZeroAddress but not sure

Reply:
Empty value for the address I think is the best null value. It also happens to be equal to ZeroAddress. Honestly not sure why they have that constant.